packages/core/index.d.ts

```diff
@@ -274,7 +275,7 @@ declare module '@rjsf/core' {
 
     export function withTheme<T = any>(
         themeProps: ThemeProps<T>,
-    ): React.ComponentClass<FormProps<T>> | React.StatelessComponent<FormProps<T>>;
+    ): typeof Form;
```

Comment:
Here's my take on this, in use in a company project for a year.
It solves the generics problem too.

For reference, it is then used in a module that has form factory filling specific needs in our project (additional default behavior, external styling additions, data type references in event handlers, error message translation, etc..).

_withTheme.tsx_
```ts
import React, { ReactElement, Ref, RefAttributes, forwardRef } from 'react';

import JsonSchemaForm, { FormProps, ThemeProps } from '@rjsf/core';

type TFormWithTheme = <T = {}>(
  p: FormProps<T> & RefAttributes<JsonSchemaForm<T>>
) => ReactElement | null;

export function withTheme<T = {}>(themeProps: ThemeProps<T>): TFormWithTheme {
  return forwardRef((props: FormProps<T>, ref: Ref<JsonSchemaForm<T>>) => {
    const { fields, widgets, ...directProps } = props;

    return (
      <JsonSchemaForm<T>
        {...themeProps}
        {...directProps}
        fields={{ ...themeProps.fields, ...fields }}
        widgets={{ ...themeProps.widgets, ...widgets }}
        ref={ref}
      />
    );
  }) as TFormWithTheme;
}
```

_usage.tsx_
```ts
const FormWithMUITheme = withTheme(MuiTheme);

type TForm = <T = {}>(
  p: FormProps<T> & RefAttributes<JsonSchemaForm<T>>
) => ReactElement | null;

export const Form: TForm = forwardRef(
    <T extends unknown = {}>(props: FormProps<T>, ref: Ref<JsonSchemaForm<T>>) => {
      const { whatever, ...otherProps } = props;
    
      /* various operations */

      return (
        <FormWithMUITheme<T>
          ref={ref}
          {...otherProps}
        />
      )}
) as TForm;
```